File: src/services/WooCommerceAPI.ts
import axios, { AxiosInstance } from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import RNFS from 'react-native-fs';
import { Book } from '../context/BookContext';

interface WooCommerceConfig {
  url: string;
  consumerKey: string;
  consumerSecret: string;
}

interface AuthResponse {
  success: boolean;
  token?: string;
  user?: any;
  message?: string;
}

interface BooksResponse {
  success: boolean;
  books?: Book[];
  message?: string;
}

interface DownloadResponse {
  success: boolean;
  file_path?: string;
  message?: string;
}

class WooCommerceAPI {
  private static instance: WooCommerceAPI;
  private api: AxiosInstance;
  private config: WooCommerceConfig | null = null;

  private constructor() {
    this.api = axios.create({
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });

    this.loadConfig();
  }

  public static getInstance(): WooCommerceAPI {
    if (!WooCommerceAPI.instance) {
      WooCommerceAPI.instance = new WooCommerceAPI();
    }
    return WooCommerceAPI.instance;
  }

  private async loadConfig() {
    try {
      const config = await AsyncStorage.getItem('woocommerce_config');
      if (config) {
        this.config = JSON.parse(config);
        this.setupAPI();
      }
    } catch (error) {
      console.error('Error loading WooCommerce config:', error);
    }
  }

  public async setConfig(config: WooCommerceConfig) {
    this.config = config;
    await AsyncStorage.setItem('woocommerce_config', JSON.stringify(config));
    this.setupAPI();
  }

  private setupAPI() {
    if (!this.config) return;

    this.api = axios.create({
      baseURL: `${this.config.url}/wp-json/wc/v3`,
      timeout: 30000,
      auth: {
        username: this.config.consumerKey,
        password: this.config.consumerSecret,
      },
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  public async authenticateUser(email: string, password: string): Promise<AuthResponse> {
    try {
      if (!this.config) {
        return { success: false, message: 'WooCommerce not configured' };
      }

      // For WooCommerce, we'll use the WordPress REST API for authentication
      const response = await axios.post(`${this.config.url}/wp-json/jwt-auth/v1/token`, {
        username: email,
        password: password,
      });

      if (response.data.token) {
        // Get user details
        const userResponse = await axios.get(`${this.config.url}/wp-json/wp/v2/users/me`, {
          headers: {
            Authorization: `Bearer ${response.data.token}`,
          },
        });

        return {
          success: true,
          token: response.data.token,
          user: {
            id: userResponse.data.id,
            email: userResponse.data.email,
            username: userResponse.data.username,
            first_name: userResponse.data.first_name,
            last_name: userResponse.data.last_name,
            avatar_url: userResponse.data.avatar_urls?.['96'],
          },
        };
      }

      return { success: false, message: 'Authentication failed' };
    } catch (error: any) {
      console.error('Authentication error:', error);
      return { 
        success: false, 
        message: error.response?.data?.message || 'Authentication failed' 
      };
    }
  }

  public async registerUser(
    email: string, 
    password: string, 
    firstName: string, 
    lastName: string
  ): Promise<AuthResponse> {
    try {
      if (!this.config) {
        return { success: false, message: 'WooCommerce not configured' };
      }

      // Register user via WordPress REST API
      const response = await axios.post(`${this.config.url}/wp-json/wp/v2/users`, {
        username: email,
        email: email,
        password: password,
        first_name: firstName,
        last_name: lastName,
      }, {
        headers: {
          Authorization: `Basic ${Buffer.from(`${this.config.consumerKey}:${this.config.consumerSecret}`).toString('base64')}`,
        },
      });

      if (response.data.id) {
        // Auto-login after registration
        return await this.authenticateUser(email, password);
      }

      return { success: false, message: 'Registration failed' };
    } catch (error: any) {
      console.error('Registration error:', error);
      return { 
        success: false, 
        message: error.response?.data?.message || 'Registration failed' 
      };
    }
  }

  public async getBooks(): Promise<BooksResponse> {
    try {
      if (!this.config) {
        return { success: false, message: 'WooCommerce not configured' };
      }

      const response = await this.api.get('/products', {
        params: {
          category: 'ebooks', // Assuming you have an 'ebooks' category
          per_page: 50,
          status: 'publish',
        },
      });

      const books: Book[] = response.data.map((product: any) => ({
        id: product.id,
        name: product.name,
        description: product.description,
        price: product.price,
        regular_price: product.regular_price,
        sale_price: product.sale_price,
        images: product.images || [],
        file_type: this.determineFileType(product),
        file_size: product.meta_data?.find((meta: any) => meta.key === '_file_size')?.value,
        pages: product.meta_data?.find((meta: any) => meta.key === '_pages')?.value,
        author: product.meta_data?.find((meta: any) => meta.key === '_author')?.value,
        publisher: product.meta_data?.find((meta: any) => meta.key === '_publisher')?.value,
        isbn: product.meta_data?.find((meta: any) => meta.key === '_isbn')?.value,
        published_date: product.meta_data?.find((meta: any) => meta.key === '_published_date')?.value,
      }));

      return { success: true, books };
    } catch (error: any) {
      console.error('Error fetching books:', error);
      return { 
        success: false, 
        message: error.response?.data?.message || 'Failed to fetch books' 
      };
    }
  }

  public async getPurchasedBooks(): Promise<BooksResponse> {
    try {
      if (!this.config) {
        return { success: false, message: 'WooCommerce not configured' };
      }

      const token = await AsyncStorage.getItem('auth_token');
      if (!token) {
        return { success: false, message: 'Not authenticated' };
      }

      // Get user's orders
      const ordersResponse = await axios.get(`${this.config.url}/wp-json/wc/v3/orders`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
        params: {
          status: 'completed',
          per_page: 50,
        },
      });

      const purchasedBookIds = new Set<number>();
      const purchasedBooks: Book[] = [];

      // Extract book IDs from completed orders
      for (const order of ordersResponse.data) {
        for (const item of order.line_items) {
          if (item.meta_data?.some((meta: any) => meta.key === '_is_ebook')) {
            purchasedBookIds.add(item.product_id);
          }
        }
      }

      // Fetch book details for purchased books
      for (const bookId of purchasedBookIds) {
        try {
          const bookResponse = await this.api.get(`/products/${bookId}`);
          const product = bookResponse.data;
          
          const book: Book = {
            id: product.id,
            name: product.name,
            description: product.description,
            price: product.price,
            regular_price: product.regular_price,
            sale_price: product.sale_price,
            images: product.images || [],
            download_url: product.meta_data?.find((meta: any) => meta.key === '_download_url')?.value,
            file_type: this.determineFileType(product),
            file_size: product.meta_data?.find((meta: any) => meta.key === '_file_size')?.value,
            pages: product.meta_data?.find((meta: any) => meta.key === '_pages')?.value,
            author: product.meta_data?.find((meta: any) => meta.key === '_author')?.value,
            publisher: product.meta_data?.find((meta: any) => meta.key === '_publisher')?.value,
            isbn: product.meta_data?.find((meta: any) => meta.key === '_isbn')?.value,
            published_date: product.meta_data?.find((meta: any) => meta.key === '_published_date')?.value,
          };

          purchasedBooks.push(book);
        } catch (error) {
          console.error(`Error fetching book ${bookId}:`, error);
        }
      }

      return { success: true, books: purchasedBooks };
    } catch (error: any) {
      console.error('Error fetching purchased books:', error);
      return { 
        success: false, 
        message: error.response?.data?.message || 'Failed to fetch purchased books' 
      };
    }
  }

  public async downloadBook(bookId: number): Promise<DownloadResponse> {
    try {
      if (!this.config) {
        return { success: false, message: 'WooCommerce not configured' };
      }

      const token = await AsyncStorage.getItem('auth_token');
      if (!token) {
        return { success: false, message: 'Not authenticated' };
      }

      // Get download URL for the book
      const response = await axios.get(`${this.config.url}/wp-json/wc/v3/products/${bookId}`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      const downloadUrl = response.data.meta_data?.find((meta: any) => meta.key === '_download_url')?.value;
      
      if (!downloadUrl) {
        return { success: false, message: 'Download URL not found' };
      }

      // Create downloads directory
      const downloadsDir = `${RNFS.DocumentDirectoryPath}/downloads`;
      await RNFS.mkdir(downloadsDir, { NSURLIsExcludedFromBackupKey: true });

      // Determine file extension
      const fileExtension = downloadUrl.includes('.pdf') ? '.pdf' : '.epub';
      const fileName = `book_${bookId}${fileExtension}`;
      const filePath = `${downloadsDir}/${fileName}`;

      // Download the file
      const downloadResult = await RNFS.downloadFile({
        fromUrl: downloadUrl,
        toFile: filePath,
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }).promise;

      if (downloadResult.statusCode === 200) {
        return { success: true, file_path: filePath };
      } else {
        return { success: false, message: 'Download failed' };
      }
    } catch (error: any) {
      console.error('Error downloading book:', error);
      return { 
        success: false, 
        message: error.response?.data?.message || 'Download failed' 
      };
    }
  }

  private determineFileType(product: any): 'pdf' | 'epub' {
    const downloadUrl = product.meta_data?.find((meta: any) => meta.key === '_download_url')?.value;
    if (downloadUrl) {
      if (downloadUrl.toLowerCase().includes('.pdf')) {
        return 'pdf';
      } else if (downloadUrl.toLowerCase().includes('.epub')) {
        return 'epub';
      }
    }
    
    // Default to PDF if we can't determine
    return 'pdf';
  }
}

export const WooCommerceAPI = WooCommerceAPI.getInstance();

File: lib/services/woocommerce_api.dart
import 'dart:convert';
import 'dart:io';
import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:path_provider/path_provider.dart';
import 'package:permission_handler/permission_handler.dart';
import '../models/book.dart';
import '../models/user.dart';

class WooCommerceAPI {
  static final WooCommerceAPI _instance = WooCommerceAPI._internal();
  factory WooCommerceAPI() => _instance;
  WooCommerceAPI._internal();

  late Dio _dio;
  String? _baseUrl;
  String? _consumerKey;
  String? _consumerSecret;
  String? _authToken;

  Future<void> initialize({
    required String baseUrl,
    required String consumerKey,
    required String consumerSecret,
  }) async {
    _baseUrl = baseUrl;
    _consumerKey = consumerKey;
    _consumerSecret = consumerSecret;

    _dio = Dio(BaseOptions(
      baseUrl: '$baseUrl/wp-json/wc/v3',
      timeout: const Duration(seconds: 30),
      headers: {
        'Content-Type': 'application/json',
      },
    ));

    // Set up basic auth for WooCommerce API
    final credentials = base64Encode(utf8.encode('$consumerKey:$consumerSecret'));
    _dio.options.headers['Authorization'] = 'Basic $credentials';

    // Load saved auth token
    final prefs = await SharedPreferences.getInstance();
    _authToken = prefs.getString('auth_token');
  }

  Future<Map<String, dynamic>> authenticateUser(String email, String password) async {
    try {
      final response = await Dio().post(
        '$_baseUrl/wp-json/jwt-auth/v1/token',
        data: {
          'username': email,
          'password': password,
        },
      );

      if (response.statusCode == 200 && response.data['token'] != null) {
        _authToken = response.data['token'];
        
        // Save token
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString('auth_token', _authToken!);

        // Get user details
        final userResponse = await Dio().get(
          '$_baseUrl/wp-json/wp/v2/users/me',
          options: Options(
            headers: {
              'Authorization': 'Bearer $_authToken',
            },
          ),
        );

        final user = User.fromJson(userResponse.data);
        
        return {
          'success': true,
          'user': user,
          'token': _authToken,
        };
      }

      return {
        'success': false,
        'message': 'Authentication failed',
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'Authentication error: ${e.toString()}',
      };
    }
  }

  Future<Map<String, dynamic>> registerUser({
    required String email,
    required String password,
    required String firstName,
    required String lastName,
  }) async {
    try {
      final response = await Dio().post(
        '$_baseUrl/wp-json/wp/v2/users',
        data: {
          'username': email,
          'email': email,
          'password': password,
          'first_name': firstName,
          'last_name': lastName,
        },
        options: Options(
          headers: {
            'Authorization': 'Basic ${base64Encode(utf8.encode('$_consumerKey:$_consumerSecret'))}',
          },
        ),
      );

      if (response.statusCode == 201) {
        // Auto-login after registration
        return await authenticateUser(email, password);
      }

      return {
        'success': false,
        'message': 'Registration failed',
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'Registration error: ${e.toString()}',
      };
    }
  }

  Future<Map<String, dynamic>> getBooks() async {
    try {
      final response = await _dio.get('/products', queryParameters: {
        'category': 'ebooks',
        'per_page': 50,
        'status': 'publish',
      });

      final books = (response.data as List)
          .map((json) => Book.fromJson(json))
          .toList();

      return {
        'success': true,
        'books': books,
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'Failed to fetch books: ${e.toString()}',
      };
    }
  }

  Future<Map<String, dynamic>> getPurchasedBooks() async {
    try {
      if (_authToken == null) {
        return {
          'success': false,
          'message': 'Not authenticated',
        };
      }

      // Get user's orders
      final ordersResponse = await Dio().get(
        '$_baseUrl/wp-json/wc/v3/orders',
        options: Options(
          headers: {
            'Authorization': 'Bearer $_authToken',
          },
        ),
        queryParameters: {
          'status': 'completed',
          'per_page': 50,
        },
      );

      final Set<int> purchasedBookIds = {};
      final List<Book> purchasedBooks = [];

      // Extract book IDs from completed orders
      for (var order in ordersResponse.data) {
        for (var item in order['line_items']) {
          if (item['meta_data']?.any((meta) => meta['key'] == '_is_ebook') == true) {
            purchasedBookIds.add(item['product_id']);
          }
        }
      }

      // Fetch book details for purchased books
      for (int bookId in purchasedBookIds) {
        try {
          final bookResponse = await _dio.get('/products/$bookId');
          final book = Book.fromJson(bookResponse.data);
          purchasedBooks.add(book.copyWith(isPurchased: true));
        } catch (e) {
          print('Error fetching book $bookId: $e');
        }
      }

      return {
        'success': true,
        'books': purchasedBooks,
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'Failed to fetch purchased books: ${e.toString()}',
      };
    }
  }

  Future<Map<String, dynamic>> downloadBook(int bookId) async {
    try {
      if (_authToken == null) {
        return {
          'success': false,
          'message': 'Not authenticated',
        };
      }

      // Get book details
      final response = await _dio.get('/products/$bookId');
      final book = Book.fromJson(response.data);
      
      if (book.downloadUrl == null) {
        return {
          'success': false,
          'message': 'Download URL not found',
        };
      }

      // Request storage permission
      final permission = await Permission.storage.request();
      if (!permission.isGranted) {
        return {
          'success': false,
          'message': 'Storage permission denied',
        };
      }

      // Create downloads directory
      final directory = await getApplicationDocumentsDirectory();
      final downloadsDir = Directory('${directory.path}/downloads');
      if (!await downloadsDir.exists()) {
        await downloadsDir.create(recursive: true);
      }

      // Determine file extension
      final fileExtension = book.fileType == BookFileType.pdf ? '.pdf' : '.epub';
      final fileName = 'book_${bookId}$fileExtension';
      final filePath = '${downloadsDir.path}/$fileName';

      // Download the file
      await Dio().download(
        book.downloadUrl!,
        filePath,
        options: Options(
          headers: {
            'Authorization': 'Bearer $_authToken',
          },
        ),
      );

      return {
        'success': true,
        'file_path': filePath,
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'Download failed: ${e.toString()}',
      };
    }
  }

  Future<void> logout() async {
    _authToken = null;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('auth_token');
  }
}